### PR TITLE
ui: add statement fingerprint to insights

### LINKED ISSUE #91665: insights: add blocking transaction fingerprint to statement details page
The insights statement details page only has the blocking id. It does not have the blocking fingerprint id. This is needs to it can be linked back to the actual query and the stats page.

Jira issue: CRDB-21374

### PR DESCRIPTION
**Pull request by j82w**
1. Removes contention events from insights. This avoids tracking and storing duplicate information.

2. Add database_name, schema_name, table_name, index_name, contending_pretty_key to crdb_internal.transaction_contention_events. This avoids needing to join with multiple tables and makes it easier for users to understand.

3. Add waiting statement info to the insights transaction details page. This way if users have a large transaction with multiple statements with multiple contention events they can see which specific statement was waited on.

4. Add blocking transaction fingerprint to the insight statement details page. The user can now see which transaction is blocking the statement ,and it has a link to the blocking transaction activity page.

https://www.loom.com/share/de37b2e3007d48c2a1e3ad39ed334e20

closes: #91665

Release note (ui change): Add waiting statement id and fingerprint to insights transaction details page. Add blocking transaction id and fingerprint to the insights statement page.

### CODE REVIEW

**Review by THardy98 — COMMENTED** [review]
Just watched the demo.

On the page listing the details of a transaction execution insight, would it be possible to link the `Statement Waiting Execution ID` to its corresponding statement execution insight page?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-:-NNql7Bd7LPPLzly6Xb6:bibrfsv)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by maryliag — COMMENTED** [review]
> Reviewed 14 of 14 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @j82w)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNql51W6UhjT3Ld4q-4):*
what happens if a user renames one of these items?
___
*[`pkg/sql/crdb_internal.go` line 6459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNqlAvjDTW9uPwfVLqy:-NNqlAvjDTW9uPwfVLqz:b-tmkhzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/72b2688eec99600d95fa2494bf7da5ea9974a102/pkg/sql/crdb_internal.go#L6459)):*
> ```Go
>     contention_duration          INTERVAL NOT NULL,
>     contending_key               BYTES NOT NULL,
> 		contending_pretty_key     	 STRING NOT NULL,
> ```

nit: spacing
___
*[`pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/insightDetailsTables.tsx` line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNqlvET89F262tS3_88:-NNqlvET89F262tS3_89:by4akvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/72b2688eec99600d95fa2494bf7da5ea9974a102/pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/insightDetailsTables.tsx#L61)):*
> ```XML
>         InsightExecEnum.STATEMENT,
>       ),
>       cell: (item: ContentionEvent) => item.waitingStmtFingerprintID,
> ```

can you make this be a link to the details page for the fingerprint?
___
*[`pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/insightDetailsTables.tsx` line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNqm5BF5oD1piQ7Eix-:-NNqm5BF5oD1piQ7Eix0:bioma5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/72b2688eec99600d95fa2494bf7da5ea9974a102/pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/insightDetailsTables.tsx#L160)):*
> ```XML
>       name: "databaseName",
>       title: insightsTableTitles.databaseName(execType),
>       cell: (item: ContentionDetails) => item.databaseName,
> ```

should we also add link for the database/table/index page? If we do, you might have to save the ids instead of the names, in case people rename and you might get a database not found of even if another get renamed with the same name,  users would get confused
___
*[`pkg/sql/crdb_internal_test.go` line 986 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNqmrtt-U1V3Z_cjrKB:-NNqmrtt-U1V3Z_cjrKC:b-tmkhzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/72b2688eec99600d95fa2494bf7da5ea9974a102/pkg/sql/crdb_internal_test.go#L986)):*
> ```Go
> 					 waiting_txn_fingerprint_id, 'hex'
> 			 ) AS waiting_txn_fingerprint_id,
>       contending_pretty_key,
> ```

nit: spacing

**Comment by j82w** [comment]
To create a link for the statement fingerprint it is required to know if it is implicit or not. It should be possible to get the information, but figured it would be better to do it in a follow up PR since this is already a large PR.

**Comment by j82w** [comment]
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNql51W6UhjT3Ld4q-4):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

what happens if a user renames one of these items?
</blockquote></details>

It gets the name when the table is generated. It's possible if the user deletes a table or index that it will be empty string.

**Comment by j82w** [comment]
I added the link in the latest iteration.

**Comment by j82w** [comment]
*[`pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/insightDetailsTables.tsx` line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNqlvET89F262tS3_88:-NNqzx3WEC8uuK1lXB9G:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/72b2688eec99600d95fa2494bf7da5ea9974a102/pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/insightDetailsTables.tsx#L61)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

I figured it would be better to do in a follow up PR. We don't have if the statement is implicit or not which is required to create the link.
</blockquote></details>

Done

**Review by maryliag — APPROVED** [review]
Nice!
:lgtm: 

> Reviewed 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-:-NNr26ex85_NmqiC9TSo:bkjg46h)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @j82w)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNql51W6UhjT3Ld4q-4):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

It gets the name when the table is generated. It's possible if the user deletes a table or index that it will be empty string.
</blockquote></details>

Ah okay!
___
*[`pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/insightDetailsTables.tsx` line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNqlvET89F262tS3_88:-NNr1usc9GN6lEh5tghr:bccpe83) ([raw file](https://github.com/cockroachdb/cockroach/blob/72b2688eec99600d95fa2494bf7da5ea9974a102/pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/insightDetailsTables.tsx#L61)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Awesome!

File: pkg/ui/workspaces/cluster-ui/src/api/txnInsightsApi.ts [hunk 1]
```diff
@@ -151,56 +151,36 @@ export type TransactionContentionEventDetails = Omit<
 >;
 
 // txnContentionDetailsQuery selects information about a specific transaction contention event.
-function txnContentionDetailsQuery(filters: TxnContentionDetailsRequest) {
-  const whereClause = getTxnContentionWhereClause(
-    ` WHERE waiting_txn_id = '${filters.txnExecutionID}'`,
-    filters,
-  );
+export function txnContentionDetailsQuery(whereClause: string) {
```
Comment (xinhaoz) [inline pkg/ui/workspaces/cluster-ui/src/api/txnInsightsApi.ts:154]: nit: can we convert this `whereClause` param to an object with filters, and build the where clause in this function?

File: pkg/ui/workspaces/cluster-ui/src/api/stmtInsightsApi.ts [hunk 2]
```diff
@@ -150,21 +142,51 @@ export function getStmtInsightsApi(
...
+    if (
+      event.contentionTime == null ||
+      event.contentionTime.asMilliseconds() <= 0
+    ) {
+      continue;
     }
 
-    if (sqlResultsAreEmpty(result)) {
-      return [];
+    if (event.statementExecutionID === "00000000-0000-0000-0000-000000000000") {
```
Comment (xinhaoz) [inline pkg/ui/workspaces/cluster-ui/src/api/stmtInsightsApi.ts:180]: Do we surface these executions?

File: pkg/ui/workspaces/cluster-ui/src/api/txnInsightsApi.ts [hunk 3]
```diff
@@ -214,58 +194,83 @@ type PartialTxnContentionDetails = Omit<
...
     execType: InsightExecEnum.TRANSACTION,
   };
 }
 
-export type TxnContentionDetailsRequest = {
-  txnExecutionID: string;
-  start?: moment.Moment;
-  end?: moment.Moment;
-};
+export async function getContentionDetailsApi(
```
Comment (xinhaoz) [inline pkg/ui/workspaces/cluster-ui/src/api/txnInsightsApi.ts:216]: Can we extract this into its own file? I'm surprised there is no circular dependency between stmtInsightsApi and txnInsightsApi

File: pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/transactionInsightDetailsOverviewTab.tsx [hunk 4]
```diff
@@ -85,17 +85,23 @@ export const TransactionInsightDetailsOverviewTab: React.FC<Props> = ({
   );
 
   const blockingExecutions: ContentionEvent[] = contentionDetails?.map(x => {
+    const stmtInsight = statements.find(
+      y => y.statementExecutionID == x.waitingStmtID,
```
Comment (xinhaoz) [inline pkg/ui/workspaces/cluster-ui/src/insights/workloadInsightDetails/transactionInsightDetailsOverviewTab.tsx:89]: nit: can we rename x and y to something a bit more descriptive, e.g. `stmt` or even just `s`

**Review by xinhaoz — COMMENTED** [review]
Nice! Just some small nits and questions.

**Review by j82w — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96872)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @maryliag and @xinhaoz)

___
*[`pkg/ui/workspaces/cluster-ui/src/api/stmtInsightsApi.ts` line 180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNqj5rMzaEHTCpm7TGs-r4-180:-NNs2O5FBzTsqVeSf8Yd:b-odk5xi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed41e3eaaeb41e47abc888d036f1a23c71c27f49/pkg/ui/workspaces/cluster-ui/src/api/stmtInsightsApi.ts#L180)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Do we surface these executions?
</blockquote></details>

It looks like they all get filtered out. I'll remove the check.
___
*[`pkg/ui/workspaces/cluster-ui/src/api/txnInsightsApi.ts` line 216 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNqj5rMzaEHTCpm7TGt-r4-216:-NNsI2iJDjvE5zMV7nEx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed41e3eaaeb41e47abc888d036f1a23c71c27f49/pkg/ui/workspaces/cluster-ui/src/api/txnInsightsApi.ts#L216)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Can we extract this into its own file? I'm surprised there is no circular dependency between stmtInsightsApi and txnInsightsApi
</blockquote></details>

Done.

File: pkg/ui/workspaces/cluster-ui/src/api/contentionApi.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  waiting_stmt_id: string;
+  waiting_stmt_fingerprint_id: string;
+  schema_name: string;
+  database_name: string;
+  table_name: string;
+  index_name: string;
+  key: string;
+};
+
+export async function GetContentionDetailsApi(
```
Comment (xinhaoz) [inline pkg/ui/workspaces/cluster-ui/src/api/contentionApi.ts:45]: nit: I don't believe we capitalize functions even if they're exported for TS. Only type names / classes / components.

**Review by xinhaoz — APPROVED** [review]

**Comment by j82w** [comment]
*[`pkg/ui/workspaces/cluster-ui/src/api/contentionApi.ts` line 45 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/96872#-NNsnXR73x_vDvt8B-nC-r5-45:-NNwc4fE3RlLvNm4reNd:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7be12b77402366d38452df3ca6d50149f1067526/pkg/ui/workspaces/cluster-ui/src/api/contentionApi.ts#L45)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

nit: I don't believe we capitalize functions even if they're exported for TS. Only type names / classes / components.
</blockquote></details>

Fixed

**Pull request merged**